In [3]:
from nltk.corpus import reuters
import nltk

nltk.download('reuters')

print(reuters.readme())
print(reuters.fileids()[150:155])
print('training files : ', len([fid for fid in reuters.fileids() if fid[:5] == 'train']))
print('testing files : ', len([fid for fid in reuters.fileids() if fid[:4] == 'test']))
print('total files : ', len(reuters.fileids()))
print(reuters.words('test/15120')[:200])
print(reuters.categories('test/15120'))

[nltk_data] Downloading package reuters to
[nltk_data]     C:\Users\lcsch\AppData\Roaming\nltk_data...



      The Reuters-21578 benchmark corpus, ApteMod version

This is a publically available version of the well-known Reuters-21578
"ApteMod" corpus for text categorization.  It has been used in
publications like these:

 * Yiming Yang and X. Liu. "A re-examination of text categorization
   methods".  1999.  Proceedings of 22nd Annual International SIGIR.
   http://citeseer.nj.nec.com/yang99reexamination.html

 * Thorsten Joachims. "Text categorization with support vector
   machines: learning with many relevant features".  1998. Proceedings
   of ECML-98, 10th European Conference on Machine Learning.
   http://citeseer.nj.nec.com/joachims98text.html

ApteMod is a collection of 10,788 documents from the Reuters financial
newswire service, partitioned into a training set with 7769 documents
and a test set with 3019 documents.  The total size of the corpus is
about 43 MB.  It is also available for download from
http://kdd.ics.uci.edu/databases/reuters21578/reuters21578.html ,
which includ

## Question 1

Il y a 10788 documents au total, dont 7769 d'entraînement et 3019 de test. Nous rechercherons les paramètres optimaux sur les documents de test.

## Question 2
Transformation en dataframes pandas

In [6]:
import pandas as pd

train_fids = [fid for fid in reuters.fileids() if fid[:5] == 'train']
test_fids = [fid for fid in reuters.fileids() if fid[:4] == 'test']

def get_text_and_category(fileid):
    fulltext = ' '.join(reuters.words(fileid))
    category = 1 if 'grain' in reuters.categories(fileid) else 0
    return fulltext, category

# Create a list of tuples containing text and category for each fileid
train_data = [(get_text_and_category(fileid)) for fileid in train_fids]
train_df = pd.DataFrame(train_data, columns=['fulltext', 'category'])

test_data = [(get_text_and_category(fileid)) for fileid in test_fids]
test_df = pd.DataFrame(test_data, columns=['fulltext', 'category'])

train_df.head()
test_df.head()

fulltext  category
0  ASIAN EXPORTERS FEAR DAMAGE FROM U . S .- JAPA...         1
1  CHINA DAILY SAYS VERMIN EAT 7 - 12 PCT GRAIN S...         0
2  JAPAN TO REVISE LONG - TERM ENERGY DEMAND DOWN...         1
3  THAI TRADE DEFICIT WIDENS IN FIRST QUARTER Tha...         0
4  INDONESIA SEES CPO PRICE RISING SHARPLY Indone...         1

## Question 3

In [18]:
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.naive_bayes import ComplementNB
from sklearn.pipeline import Pipeline
from pprint import pprint
from sklearn.model_selection import GridSearchCV
from time import time

pipeline = Pipeline(
    [
        ("vect", TfidfVectorizer()),
        ("clf", ComplementNB()),
    ]
)
pipeline

import numpy as np

parameter_grid = {
    #"vect__max_df": (0.2, 0.4, 0.6, 0.8, 1.0),
    "vect__min_df": (1, 3, 5, 10),
    #"vect__ngram_range": ((1, 1), (1, 2)),  # unigrams or bigrams
    #"vect__norm": ("l1", "l2"),
    #"clf__alpha": np.logspace(-6, 6, 13),
}

random_search = GridSearchCV(
    scoring='f1',
    estimator=pipeline,
    param_grid=parameter_grid,
    n_jobs=2,
    verbose=1,
)

print("Performing grid search...")
print("Hyperparameters to be evaluated:")
pprint(parameter_grid)

t0 = time()
random_search.fit(train_df['fulltext'], train_df['category'])
print(f"Done in {time() - t0:.3f}s")

Performing grid search...
Hyperparameters to be evaluated:
{'vect__min_df': (1, 3, 5, 10)}
Fitting 5 folds for each of 4 candidates, totalling 20 fits
Done in 8.317s


In [20]:
from sklearn.metrics import f1_score

print("Best parameters combination found:")
best_parameters = random_search.best_estimator_.get_params()
for param_name in sorted(parameter_grid.keys()):
    print(f"{param_name}: {best_parameters[param_name]}")

score = random_search.score(test_df['fulltext'], test_df['category'])
print(
    "F1-score of the best parameters using the inner CV of "
    f"the random search: {random_search.best_score_:.3f}"
)
print(f"F1-score on test set: {score:.3f}")

Best parameters combination found:
vect__min_df: 5
F1-score of the best parameters using the inner CV of the random search: 0.986
F1-score on test set: 0.980


## Autres options et résultats obtenus

### Alternatives à CountVectorizer et TfidfTransformer


### Modèle utilisé

ComplementNB :
SGDClassifier :

TODO


In [ ]:
import pandas as pd


def shorten_param(param_name):
    """Remove components' prefixes in param_name."""
    if "__" in param_name:
        return param_name.rsplit("__", 1)[1]
    return param_name


cv_results = pd.DataFrame(random_search.cv_results_)
cv_results = cv_results.rename(shorten_param, axis=1)

We can use a [plotly.express.scatter](https://plotly.com/python-api-reference/generated/plotly.express.scatter.html)
to visualize the trade-off between scoring time and mean test score (i.e. "CV
score"). Passing the cursor over a given point displays the corresponding
parameters. Error bars correspond to one standard deviation as computed in the
different folds of the cross-validation.



In [ ]:
import plotly.express as px

param_names = [shorten_param(name) for name in parameter_grid.keys()]
labels = {
    "mean_score_time": "CV Score time (s)",
    "mean_test_score": "CV score (accuracy)",
}
fig = px.scatter(
    cv_results,
    x="mean_score_time",
    y="mean_test_score",
    error_x="std_score_time",
    error_y="std_test_score",
    hover_data=param_names,
    labels=labels,
)
fig.update_layout(
    title={
        "text": "trade-off between scoring time and mean test score",
        "y": 0.95,
        "x": 0.5,
        "xanchor": "center",
        "yanchor": "top",
    }
)
fig

Notice that the cluster of models in the upper-left corner of the plot have
the best trade-off between accuracy and scoring time. In this case, using
bigrams increases the required scoring time without improving considerably the
accuracy of the pipeline.

<div class="alert alert-info"><h4>Note</h4><p>For more information on how to customize an automated tuning to
   maximize score and minimize scoring time, see the example notebook
   `sphx_glr_auto_examples_model_selection_plot_grid_search_digits.py`.</p></div>

We can also use a [plotly.express.parallel_coordinates](https://plotly.com/python-api-reference/generated/plotly.express.parallel_coordinates.html)
to further visualize the mean test score as a function of the tuned
hyperparameters. This helps finding interactions between more than two
hyperparameters and provide intuition on their relevance for improving the
performance of a pipeline.

We apply a `math.log10` transformation on the `alpha` axis to spread the
active range and improve the readability of the plot. A value $x$ on
said axis is to be understood as $10^x$.



In [ ]:
import math

column_results = param_names + ["mean_test_score", "mean_score_time"]

transform_funcs = dict.fromkeys(column_results, lambda x: x)
# Using a logarithmic scale for alpha
transform_funcs["alpha"] = math.log10
# L1 norms are mapped to index 1, and L2 norms to index 2
transform_funcs["norm"] = lambda x: 2 if x == "l2" else 1
# Unigrams are mapped to index 1 and bigrams to index 2
transform_funcs["ngram_range"] = lambda x: x[1]

fig = px.parallel_coordinates(
    cv_results[column_results].apply(transform_funcs),
    color="mean_test_score",
    color_continuous_scale=px.colors.sequential.Viridis_r,
    labels=labels,
)
fig.update_layout(
    title={
        "text": "Parallel coordinates plot of text classifier pipeline",
        "y": 0.99,
        "x": 0.5,
        "xanchor": "center",
        "yanchor": "top",
    }
)
fig

The parallel coordinates plot displays the values of the hyperparameters on
different columns while the performance metric is color coded. It is possible
to select a range of results by clicking and holding on any axis of the
parallel coordinate plot. You can then slide (move) the range selection and
cross two selections to see the intersections. You can undo a selection by
clicking once again on the same axis.

In particular for this hyperparameter search, it is interesting to notice that
the top performing models do not seem to depend on the regularization `norm`,
but they do depend on a trade-off between `max_df`, `min_df` and the
regularization strength `alpha`. The reason is that including noisy features
(i.e. `max_df` close to $1.0$ or `min_df` close to $0$) tend to
overfit and therefore require a stronger regularization to compensate. Having
less features require less regularization and less scoring time.

The best accuracy scores are obtained when `alpha` is between $10^{-6}$
and $10^0$, regardless of the hyperparameter `norm`.

